windows/go-1.11 compile error

I am seeing the following error trying to use this package on Windows using go-1.11beta1:

> 13:11:05 ....\vendor\github.com\google\certificate-transparency\go\x509\root_windows.go:112:3: cannot use uintptr(unsafe.Pointer(sslPara)) (type uintptr) as type syscall.Pointer in field value

This is apparently caused by the following change in golang: https://github.com/golang/go/commit/4869ec00e87ef49db2646c25d28d5c7e4f1caff8

and requires a one-line fix identical to the one in here: https://github.com/golang/go/commit/4869ec00e87ef49db2646c25d28d5c7e4f1caff8?diff=unified#diff-9a0dd81b66ea632b270ed4c40cb71e43

which of course will break compilation with golang < 1.11, so a more complicated fix is required.

Found while trying to compile Docker with go-1.11beta1 on Windows (https://github.com/moby/moby/pull/37358#issuecomment-401923825)

--- a/x509/ptr_sysptr_windows.go
+++ b/x509/ptr_sysptr_windows.go
@@ -1,0 +1,20 @@
+// Copyright 2018 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build go1.11
+
+package x509
+
+import (
+	"syscall"
+	"unsafe"
+)
+
+// For Go versions >= 1.11, the ExtraPolicyPara field in
+// syscall.CertChainPolicyPara is of type syscall.Pointer.  See:
+//   https://github.com/golang/go/commit/4869ec00e87ef
+
+func convertToPolicyParaType(p unsafe.Pointer) syscall.Pointer {
+	return (syscall.Pointer)(p)
+}

--- a/x509/ptr_uint_windows.go
+++ b/x509/ptr_uint_windows.go
@@ -1,0 +1,17 @@
+// Copyright 2018 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build !go1.11
+
+package x509
+
+import "unsafe"
+
+// For Go versions before 1.11, the ExtraPolicyPara field in
+// syscall.CertChainPolicyPara was of type uintptr.  See:
+//   https://github.com/golang/go/commit/4869ec00e87ef
+
+func convertToPolicyParaType(p unsafe.Pointer) uintptr {
+	return uintptr(p)
+}

--- a/x509/root_windows.go
+++ b/x509/root_windows.go
@@ -109,7 +109,7 @@ func checkChainSSLServerPolicy(c *Certificate, chainCtx *syscall.CertChainContex
 	sslPara.Size = uint32(unsafe.Sizeof(*sslPara))
 
 	para := &syscall.CertChainPolicyPara{
-		ExtraPolicyPara: uintptr(unsafe.Pointer(sslPara)),
+		ExtraPolicyPara: convertToPolicyParaType(unsafe.Pointer(sslPara)),
 	}
 	para.Size = uint32(unsafe.Sizeof(*para))
 

